leopard.go

```diff
@@ -1,54 +1,72 @@
-//go:build leopard
-
-// Note that if the build tag leopard is used, liblibleopard.a
-// has to be present where the linker will find it.
-// Otherwise go-leopard won't build.
 package rsmt2d
 
-import "github.com/celestiaorg/go-leopard"
+import (
+	"sync"
 
-var _ Codec = leoRSFF8Codec{}
-var _ Codec = leoRSFF16Codec{}
+	"github.com/klauspost/reedsolomon"
+)
 
-func init() {
-	registerCodec(LeopardFF8, newLeoRSFF8Codec())
-	registerCodec(LeopardFF16, newLeoRSFF16Codec())
-}
+var _ Codec = leoRSCodec{}
 
-type leoRSFF8Codec struct{}
-
-func (l leoRSFF8Codec) Encode(data [][]byte) ([][]byte, error) {
-	return leopard.Encode(data)
-}
-
-func (l leoRSFF8Codec) Decode(data [][]byte) ([][]byte, error) {
-	half := len(data) / 2
-	return leopard.Decode(data[:half], data[half:])
+func init() {
+	registerCodec(Leopard, NewLeoRSCodec())
 }
 
-func (l leoRSFF8Codec) maxChunks() int {
-	return 128 * 128
+type leoRSCodec struct {
+	// Cache the encoders of various sizes to not have to re-instantiate those
+	// as it is costly.
+	//
+	// Note that past sizes are not removed from the cache at all as the various
+	// data sizes are expected to relatively small and will not cause any memory issue.
+	//
+	// TODO: switch to a generic version of sync.Map with type reedsolomon.Encoder
+	// once it made it into the standard lib
+	encCache sync.Map
 }
 
-func newLeoRSFF8Codec() leoRSFF8Codec {
-	return leoRSFF8Codec{}
-}
+func (l leoRSCodec) Encode(data [][]byte) ([][]byte, error) {
```

Comment:
blocking bug: we should use a pointer here! Otherwise, we copy `leoRSCodec` and the whole `sync.Map`, while it [must not be copied](https://pkg.go.dev/sync#Map).

Reply:
Makes sense! Thanks for catching this 🙏🏼

Reply:
Fixed in [a424318](https://github.com/celestiaorg/rsmt2d/pull/136/commits/a424318e88877896bd1fa15e4b5d893c07f9a779)

Reply:
Good catch! I wasn't aware of this